### PR TITLE
Add ProgressMeter.jl support and allow parallelized objective functions

### PR DESCRIPTION
**Pull request by atrophiedbrain**

### CODE REVIEW

**Comment by tpapp** [comment]
Thanks for the PR. I will review it but currently I am very busy. Feel free to ping me in a week if I don't get to it.

**Comment by atrophiedbrain** [comment]
@tpapp any issues I should address?

File: test/runtests.jl [hunk 1]
```diff
@@ -13,14 +13,18 @@ include("test_functions.jl")
 end
 
 @testset "global optimization" begin
+    pr = true
```
Comment (tpapp) [inline test/runtests.jl:16]: Please test for all combinations (eg two outer for loops for `pr = [true, false]` etc).

File: test/runtests.jl [hunk 2]
```diff
@@ -13,14 +13,18 @@ include("test_functions.jl")
...
+    psp = false
     for F in setdiff(TEST_FUNCTIONS, (RASTRIGIN, )) # Rastrigin disabled for now
         n = 10
         P = MinimizationProblem(F, lower_bounds(F, n), upper_bounds(F, n))
         local_method = NLoptLocalMethod(NLopt.LN_BOBYQA)
         multistart_method = TikTak(100)
-        p = multistart_minimization(multistart_method, local_method, P)
+        p = multistart_minimization(multistart_method, local_method, P; progress=pr, parallel_sobol_points=psp)
```
Comment (tpapp) [inline test/runtests.jl:23]:
```suggestion
        p = multistart_minimization(multistart_method, local_method, P; progress = pr, parallel_sobol_points = psp)
```

File: test/runtests.jl [hunk 3]
```diff
@@ -13,14 +13,18 @@ include("test_functions.jl")
...
         x₀ = minimum_location(F, n)
         @test p.location ≈ x₀ atol = 1e-5
         @test p.value ≈ F(x₀) atol = 1e-10
+        pr = false
+        psp = true
```
Comment (tpapp) [inline test/runtests.jl:28]:
```suggestion
```

File: src/MultistartOptimization.jl [hunk 4]
```diff
@@ -178,17 +183,25 @@ $(SIGNATURES)
...
         local_minimum = local_minimization(local_method, minimization_problem, x)
+        if @isdefined(prog)
+            ProgressMeter.next!(prog)
+        end
         local_minimum.value < visited_minimum.value ? local_minimum : visited_minimum
     end
     foldl(_step, enumerate(initial_points); init = first(initial_points))
 end
 
+
```
Comment (tpapp) [inline src/MultistartOptimization.jl:206]:
```suggestion
```
(removed dangling space)

File: src/MultistartOptimization.jl [hunk 5]
```diff
@@ -60,11 +61,15 @@ Evaluate and return points of an `N`-element Sobol sequence.
...
+function sobol_starting_points(minimization_problem::MinimizationProblem, N::Integer; parallel_map=true)
     @unpack objective, lower_bounds, upper_bounds = minimization_problem
     s = SobolSeq(lower_bounds, upper_bounds)
     skip(s, N)                  # better uniformity
-    map(fetch, map(x -> @spawn(LocationValue(x, objective(x))), Iterators.take(s, N)))
+    if parallel_map
+        map(fetch, map(x -> @spawn(LocationValue(x, objective(x))), Iterators.take(s, N)))
+    else
+        map(x -> LocationValue(x, objective(x)), Iterators.take(s, N))
+    end
```
Comment (tpapp) [inline src/MultistartOptimization.jl:72]: DRY, write something like
```julia
f(x) = @spawn(LocationValue(x, objective(x)))
```
and reuse it below.

File: src/MultistartOptimization.jl [hunk 6]
```diff
@@ -60,11 +61,15 @@ Evaluate and return points of an `N`-element Sobol sequence.
 
 An effort is made to parallelize the code using `Threads` when available.
 """
-function sobol_starting_points(minimization_problem::MinimizationProblem, N::Integer)
+function sobol_starting_points(minimization_problem::MinimizationProblem, N::Integer; parallel_map=true)
```
Comment (tpapp) [inline src/MultistartOptimization.jl:64]:
```suggestion
function sobol_starting_points(minimization_problem::MinimizationProblem, N::Integer; parallel_map = true)
```

File: src/MultistartOptimization.jl [hunk 7]
```diff
@@ -60,11 +61,15 @@ Evaluate and return points of an `N`-element Sobol sequence.
 
 An effort is made to parallelize the code using `Threads` when available.
```
Comment (tpapp) [inline src/MultistartOptimization.jl:62]: document the `parallel_map` keyword argument (start a `# Keyword arguments` part)

**Review by tpapp — CHANGES_REQUESTED** [review]
LGTM, just requested some style changes.

**Comment by tpapp** [comment]
Thanks for the ping, I just reviewed and requested some minor changes.

**Comment by tpapp** [comment]
@atrophiedbrain, this is a very nice PR, are you still interested in finishing it?

**Comment by atrophiedbrain** [comment]
I am, but school has me very preoccupied at the moment so it may be a while...

**Comment by tpapp** [comment]
Perhaps a more generic way would be using [Folds.jl](https://discourse.julialang.org/t/ann-folds-jl-threaded-distributed-and-gpu-based-high-level-data-parallel-interface-for-julia/54701), which would allow the user to choose between threads and other implementations.

**Pull request closed without merging**